File: go/size_test.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

package bnf

// The size of what the emitter produces is a contract (tabnas/bnf#71).
// Mirrors ts/test/size.test.js: a choice is dispatched on the first token
// of each alternative, and the lookahead deepens only under a head two
// alternatives share, so the emitted table grows with the number of
// decisions rather than with the product of the tokens that can fill
// four positions. Every count is exact and pinned.

import (
	"encoding/json"
	"fmt"
	"reflect"
	"regexp"
	"strings"
	"testing"

	tabnas "github.com/tabnas/parser/go"
)

func szLit(s string) *Element {
	return &Element{Kind: KindTerm, Literal: s, CaseSensitive: true, HasCaseSens: true}
}

func szKw(n int) *Production {
	alts := []Sequence{}
	for i := 1; i <= n; i++ {
		alts = append(alts, Sequence{szLit(fmt.Sprintf("k%d", i))})
	}
	return &Production{Name: "kw", Alts: alts}
}

// doc = "{" *entry "}" ; entry = kw kw ; kw = "k1" / ... / "kN"
func szStarred(n int) *Grammar {
	return &Grammar{Productions: []*Production{
		{Name: "doc", Alts: []Sequence{{szLit("{"), {Kind: KindStar, Inner: ref("entry")}, szLit("}")}}},
		{Name: "entry", Alts: []Sequence{{ref("kw"), ref("kw")}}},
		szKw(n),
	}}
}

// doc = "{" x "}" ; x = entry entry / "z" ; entry = kw kw ; kw = ...
func szChoice(n int) *Grammar {
	return &Grammar{Productions: []*Production{
		{Name: "doc", Alts: []Sequence{{szLit("{"), ref("x"), szLit("}")}}},
		{Name: "x", Alts: []Sequence{{ref("entry"), ref("entry")}, {szLit("z")}}},
		{Name: "entry", Alts: []Sequence{{ref("kw"), ref("kw")}}},
		szKw(n),
	}}
}

func szOpens(t *testing.T, spec *tabnas.GrammarSpec, name string) int {
	t.Helper()
	r, ok := spec.Rule[name]
	if !ok || r == nil {
		t.Fatalf("no rule %q", name)
	}
	v := reflect.ValueOf(r.Open)
	if v.IsValid() && v.Kind() == reflect.Slice {
		return v.Len()
	}
	t.Fatalf("rule %q: open is %T", name, r.Open)
	return 0
}

func szRule(t *testing.T, spec *tabnas.GrammarSpec, re string) string {
	t.Helper()
	rx := regexp.MustCompile(re)
	for name := range spec.Rule {
		if rx.MatchString(name) {
			return name
		}
	}
	t.Fatalf("no rule matching %s", re)
	return ""
}

func szTotal(t *testing.T, spec *tabnas.GrammarSpec) int {
	t.Helper()
	total := 0
	for name, r := range spec.Rule {
		if r == nil {
			continue
		}
		total += szOpens(t, spec, name)
	}
	return total
}

func szParses(t *testing.T, spec *tabnas.GrammarSpec, src string) bool {
	t.Helper()
	j := tabnas.Make()
	if err := j.Grammar(spec); err != nil {
		t.Fatalf("install: %v", err)
	}
	_, err := j.Parse(src)
	return err == nil
}

func TestSizeRepeatedEntryDispatchesOnHeads(t *testing.T) {
	// The star helper: one entry per keyword head, one FOLLOW peek for
	// the `}` that ends the loop, and the bare fallback.
	for _, n := range []int{2, 8, 26} {
		spec, err := EmitGrammarSpec(szStarred(n), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true})
		if err != nil {
			t.Fatal(err)
		}
		helper := szRule(t, spec, `star_entry$`)
		if got := szOpens(t, spec, helper); got != n+2 {
			t.Errorf("N=%d: %s has %d open alternates, want %d", n, helper, got, n+2)
		}
	}
}

// A leading reference to the class is consumed as its one token where the
// plain compile inlines the class's alternatives, and stays a node where
// the plain compile keeps the reference: the same parse result, node for
// node, with the option on or off.
func TestSizeTokenClassLeavesTheTreeAsTheOptionOffCompileDoes(t *testing.T) {
	off, err := EmitGrammarSpec(szStarred(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true})
	if err != nil {
		t.Fatal(err)
	}
	on, err := EmitGrammarSpec(szStarred(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true, TokenClasses: true})
	if err != nil {
		t.Fatal(err)
	}
	joff, jon := tabnas.Make(), tabnas.Make()
	if err := joff.Grammar(off); err != nil {
		t.Fatal(err)
	}
	if err := jon.Grammar(on); err != nil {
		t.Fatal(err)
	}
	for _, src := range []string{"{k1 k2 k26 k1}", "{}", "{k3 k3}"} {
		a, err := joff.Parse(src)
		if err != nil {
			t.Fatalf("%s: off: %v", src, err)
		}
		b, err := jon.Parse(src)
		if err != nil {
			t.Fatalf("%s: on: %v", src, err)
		}
		ja, _ := json.Marshal(a)
		jb, _ := json.Marshal(b)
		if string(ja) != string(jb) {
			t.Fatalf("%s: trees differ\n off: %s\n on:  %s", src, ja, jb)
		}
	}
}

func TestSizeUncontestedChoiceDispatchesOnOneToken(t *testing.T) {
	spec, err := EmitGrammarSpec(szChoice(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true})
	if err != nil {
		t.Fatal(err)
	}
	// 26 keyword heads for the first alternative, one for "z".
	if got := szOpens(t, spec, "x"); got != 27 {
		t.Errorf("x has %d open alternates, want 27", got)
	}
}

func TestSizeTokenClassIsOneLookaheadToken(t *testing.T) {
	spec, err := EmitGrammarSpec(szStarred(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true, TokenClasses: true})
	if err != nil {
		t.Fatal(err)
	}
	// The class is a set; the helper peeks it once. The class rule
	// itself keeps its 26 alternates, one per member: it is the rule
	// that builds the node.
	if len(spec.Options.TokenSet) != 1 || len(spec.Options.TokenSet["kw"]) != 26 {
		t.Fatalf("token sets: %v", spec.Options.TokenSet)
	}
	helper := szRule(t, spec, `star_entry$`)
	if got := szOpens(t, spec, helper); got != 3 {
		t.Errorf("%s has %d open alternates, want 3", helper, got)
	}
	if got := szOpens(t, spec, "kw"); got != 26 {
		t.Errorf("kw has %d open alternates, want 26", got)
	}
	x, err := EmitGrammarSpec(szChoice(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true, TokenClasses: true})
	if err != nil {
		t.Fatal(err)
	}
	if got := szOpens(t, x, "x"); got != 2 {
		t.Errorf("x has %d open alternates, want 2", got)
	}
	// The whole grammar stays small, and installs and parses.
	if total := szTotal(t, spec); total >= 40 {
		t.Errorf("%d open alternates", total)
	}
	j := tabnas.Make()
	if err := j.Grammar(spec); err != nil {
		t.Fatal(err)
	}
	out, err := j.Parse("{k1 k2 k26 k1}")
	if err != nil {
		t.Fatal(err)
	}
	// The class keeps its node: it is a rule of its own, not inlined.
	if !strings.Contains(fmt.Sprintf("%v", out), "kw") {
		t.Errorf("no kw node in %v", out)
	}
	if _, err := j.Parse("{k1}"); err == nil {
		t.Error("{k1} should be refused")
	}
}

func TestSizeEmittedGrammarInstallsAndParsesAtN26(t *testing.T) {
	spec, err := EmitGrammarSpec(szStarred(26), &ConvertOptions{Tag: "rp", Start: "doc", WordKeywords: true})
	if err != nil {
		t.Fatal(err)
	}
	if !szParses(t, spec, "{k1 k2 k2 k1}") {
		t.Error("{k1 k2 k2 k1} should parse")
	}
	if szParses(t, spec, "{k1}") {
		t.Error("{k1} should be refused")
	}
}

func TestSizeContestedHeadDeepensOnlyAsFarAsNeeded(t *testing.T) {
	// Two alternatives share `a`; they part at the second token, so the
	// dispatcher peeks two tokens under `a` and one under `z`.
	spec, err := EmitGrammarSpec(&Grammar{Productions: []*Production{
		{Name: "doc", Alts: []Sequence{{ref("x")}}},
		{Name: "x", Alts: []Sequence{
			{szLit("a"), ref("t"), ref("u")},
			{szLit("a"), ref("u"), ref("t")},
			{szLit("z"), ref("t"), ref("u")},
		}},
		{Name: "t", Alts: []Sequence{{szLit("."), szLit(".")}}},
		{Name: "u", Alts: []Sequence{{szLit(","), szLit(",")}}},
	}}, &ConvertOptions{Tag: "rp", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	opens := reflect.ValueOf(spec.Rule["x"].Open)
	got := []string{}
	for i := 0; i < opens.Len(); i++ {
		o := reflect.Indirect(opens.Index(i))
		got = append(got, fmt.Sprintf("%v|%v", o.FieldByName("S"), o.FieldByName("B")))
	}
	want := []string{"#A #T|2", "#A #T1|2", "#Z|1"}
	if fmt.Sprint(got) != fmt.Sprint(want) {
		t.Errorf("x dispatch: %v, want %v", got, want)
	}
	for _, src := range []string{"a..,,", "a,,..", "z..,,"} {
		if !szParses(t, spec, src) {
			t.Errorf("%q should parse", src)
		}
	}
	if szParses(t, spec, "a.,.,") {
		t.Error("a.,., should be refused")
	}
}

func TestSizeRepetitionContestedByFollowLooksAcrossTheBoundary(t *testing.T) {
	// start = *( "a" t ) "a" "b" with t = "x" "x": at an `a` the loop may
	// continue (`a x`) or exit (`a b`). The exit is an open path, so the
	// continue entries keep the full window here, as they always did.
	spec, err := EmitGrammarSpec(&Grammar{Productions: []*Production{
		{Name: "start", Alts: []Sequence{{
			{Kind: KindStar, Inner: groupEl(Sequence{szLit("a"), ref("t")})},
			szLit("a"), szLit("b"),
		}}},
		{Name: "t", Alts: []Sequence{{szLit("x"), szLit("x")}}},
	}}, &ConvertOptions{Tag: "rp", Start: "start"})
	if err != nil {
		t.Fatal(err)
	}
	helper := szRule(t, spec, `star.*group$`)
	opens := reflect.ValueOf(spec.Rule[helper].Open)
	first := reflect.Indirect(opens.Index(0))
	if s := fmt.Sprint(first.FieldByName("S")); !strings.HasPrefix(s, "#A #X") {
		t.Errorf("first continue entry peeks %q", s)
	}
	for _, src := range []string{"ab", "axxab", "axxaxxab"} {
		if !szParses(t, spec, src) {
			t.Errorf("%q should parse", src)
		}
	}
	for _, src := range []string{"b", "axx"} {
		if szParses(t, spec, src) {
			t.Errorf("%q should be refused", src)
		}
	}
}

File: rs/tests/size_test.rs
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

//! The size of what the emitter produces is a contract (tabnas/bnf#71).
//! Mirrors `ts/test/size.test.js` and `go/size_test.go`: a choice is
//! dispatched on the first token of each alternative, and the lookahead
//! deepens only under a head two alternatives share, so the emitted table
//! grows with the number of decisions rather than with the product of the
//! tokens that can fill four positions. Every count is exact and pinned.

mod common;

use common::{group, prod, reference, sens_term, star};
use tabnas_bnf::{emit_grammar_spec, ConvertOptions, Element, Grammar, GrammarSpec, Production};

fn kw(n: usize) -> Production {
    let alts: Vec<Vec<Element>> = (1..=n).map(|i| vec![sens_term(&format!("k{i}"))]).collect();
    prod("kw", alts)
}

// doc = "{" *entry "}" ; entry = kw kw ; kw = "k1" / ... / "kN"
fn starred(n: usize) -> Grammar {
    Grammar::new(vec![
        prod(
            "doc",
            vec![vec![
                sens_term("{"),
                star(reference("entry")),
                sens_term("}"),
            ]],
        ),
        prod("entry", vec![vec![reference("kw"), reference("kw")]]),
        kw(n),
    ])
}

// doc = "{" x "}" ; x = entry entry / "z" ; entry = kw kw ; kw = ...
fn choice(n: usize) -> Grammar {
    Grammar::new(vec![
        prod(
            "doc",
            vec![vec![sens_term("{"), reference("x"), sens_term("}")]],
        ),
        prod(
            "x",
            vec![
                vec![reference("entry"), reference("entry")],
                vec![sens_term("z")],
            ],
        ),
        prod("entry", vec![vec![reference("kw"), reference("kw")]]),
        kw(n),
    ])
}

fn opens(spec: &GrammarSpec, name: &str) -> usize {
    spec.rule
        .get(name)
        .and_then(|r| r.as_ref())
        .map(|r| r.open.len())
        .unwrap_or_else(|| panic!("no rule {name}"))
}

fn rule_matching(spec: &GrammarSpec, f: impl Fn(&str) -> bool) -> String {
    spec.rule
        .keys()
        .find(|n| f(n))
        .cloned()
        .unwrap_or_else(|| panic!("no matching rule"))
}

fn total_opens(spec: &GrammarSpec) -> usize {
    spec.rule
        .values()
        .filter_map(|r| r.as_ref())
        .map(|r| r.open.len())
        .sum()
}

fn parses(spec: &GrammarSpec, src: &str) -> bool {
    let mut parser = tabnas::Tabnas::new();
    spec.install(&mut parser).expect("install");
    parser.parse(src).is_ok()
}

fn opts() -> ConvertOptions {
    ConvertOptions::tag("rp").start("doc").word_keywords(true)
}

#[test]
fn a_repeated_entry_dispatches_on_heads_not_paths() {
    // The star helper: one entry per keyword head, one FOLLOW peek for
    // the `}` that ends the loop, and the bare fallback.
    for n in [2usize, 8, 26] {
        let spec = emit_grammar_spec(&starred(n), &opts()).expect("emit");
        let helper = rule_matching(&spec, |name| name.ends_with("star_entry"));
        assert_eq!(opens(&spec, &helper), n + 2, "N={n}: {helper}");
    }
}

#[test]
fn an_uncontested_choice_dispatches_on_one_token_per_head() {
    let spec = emit_grammar_spec(&choice(26), &opts()).expect("emit");
    // 26 keyword heads for the first alternative, one for "z".
    assert_eq!(opens(&spec, "x"), 27);
}

#[test]
fn a_token_class_is_one_lookahead_token() {
    let spec = emit_grammar_spec(&starred(26), &opts().token_classes(true)).expect("emit");
    // The class is a set; the helper peeks it once. The class rule
    // itself keeps its 26 alternates, one per member: it is the rule
    // that builds the node.
    let sets = spec.options.get("tokenSet").expect("token sets");
    let kw_set = sets.get("kw").expect("kw set");
    assert_eq!(kw_set.as_array().map(|a| a.len()), Some(26));
    assert_eq!(sets.as_object().map(|o| o.len()), Some(1));
    let helper = rule_matching(&spec, |name| name.ends_with("star_entry"));
    assert_eq!(opens(&spec, &helper), 3, "{helper}");
    assert_eq!(opens(&spec, "kw"), 26);
    let x = emit_grammar_spec(&choice(26), &opts().token_classes(true)).expect("emit");
    assert_eq!(opens(&x, "x"), 2);
    // The whole grammar stays small, and installs and parses.
    assert!(
        total_opens(&spec) < 40,
        "{} open alternates",
        total_opens(&spec)
    );
    let mut parser = tabnas::Tabnas::new();
    spec.install(&mut parser).expect("install");
    let out = parser.parse("{k1 k2 k26 k1}").expect("parse");
    // The class keeps its node: it is a rule of its own, not inlined.
    assert!(out.to_json().to_string().contains("\"kw\""));
    assert!(parser.parse("{k1}").is_err());
}

#[test]
fn a_token_class_leaves_the_tree_as_the_option_off_compile_does() {
    // A leading reference to the class is consumed as its one token where
    // the plain compile inlines the class's alternatives, and stays a
    // node where the plain compile keeps the reference: the same parse
    // result, node for node, with the option on or off.
    let off = emit_grammar_spec(&starred(26), &opts()).expect("emit");
    let on = emit_grammar_spec(&starred(26), &opts().token_classes(true)).expect("emit");
    let mut poff = tabnas::Tabnas::new();
    off.install(&mut poff).expect("install");
    let mut pon = tabnas::Tabnas::new();
    on.install(&mut pon).expect("install");
    for src in ["{k1 k2 k26 k1}", "{}", "{k3 k3}"] {
        let a = poff.parse(src).expect("off");
        let b = pon.parse(src).expect("on");
        assert_eq!(a.to_json(), b.to_json(), "{src}");
    }
}

#[test]
fn the_emitted_grammar_installs_and_parses_at_n26_either_way() {
    let spec = emit_grammar_spec(&starred(26), &opts()).expect("emit");
    assert!(parses(&spec, "{k1 k2 k2 k1}"));
    assert!(!parses(&spec, "{k1}"));
}

#[test]
fn a_contested_head_deepens_only_as_far_as_the_decision_needs() {
    // Two alternatives share `a`; they part at the second token, so the
    // dispatcher peeks two tokens under `a` and one under `z`.
    let spec = emit_grammar_spec(
        &Grammar::new(vec![
            prod("doc", vec![vec![reference("x")]]),
            prod(
                "x",
                vec![
                    vec![sens_term("a"), reference("t"), reference("u")],
                    vec![sens_term("a"), reference("u"), reference("t")],
                    vec![sens_term("z"), reference("t"), reference("u")],
                ],
            ),
            prod("t", vec![vec![sens_term("."), sens_term(".")]]),
            prod("u", vec![vec![sens_term(","), sens_term(",")]]),
        ]),
        &ConvertOptions::tag("rp").start("doc"),
    )
    .expect("emit");
    let x = spec.rule.get("x").and_then(|r| r.as_ref()).expect("x");
    let got: Vec<(String, u64)> = x
        .open
        .iter()
        .map(|o| {
            (
                o.s().unwrap_or("").to_string(),
                o.get("b").and_then(|b| b.as_u64()).unwrap_or(0),
            )
        })
        .collect();
    assert_eq!(
        got,
        vec![
            ("#A #T".to_string(), 2),
            ("#A #T1".to_string(), 2),
            ("#Z".to_string(), 1)
        ]
    );
    for src in ["a..,,", "a,,..", "z..,,"] {
        assert!(parses(&spec, src), "{src} should parse");
    }
    assert!(!parses(&spec, "a.,.,"));
}

#[test]
fn a_repetition_contested_by_what_follows_it_looks_across_the_boundary() {
    // start = *( "a" t ) "a" "b" with t = "x" "x": at an `a` the loop may
    // continue (`a x`) or exit (`a b`). The exit is an open path, so the
    // continue entries keep the full window here, as they always did.
    let spec = emit_grammar_spec(
        &Grammar::new(vec![
            prod(
                "start",
                vec![vec![
                    star(group(vec![vec![sens_term("a"), reference("t")]])),
                    sens_term("a"),
                    sens_term("b"),
                ]],
            ),
            prod("t", vec![vec![sens_term("x"), sens_term("x")]]),
        ]),
        &ConvertOptions::tag("rp").start("start"),
    )
    .expect("emit");
    let helper = rule_matching(&spec, |name| {
        name.contains("star") && name.ends_with("group")
    });
    let first = spec.rule[&helper].as_ref().expect("helper").open[0]
        .s()
        .unwrap_or("")
        .to_string();
    assert!(first.starts_with("#A #X"), "{first}");
    for src in ["ab", "axxab", "axxaxxab"] {
        assert!(parses(&spec, src), "{src} should parse");
    }
    for src in ["b", "axx"] {
        assert!(!parses(&spec, src), "{src} should be refused");
    }
}
